### ISSUE TITLE
Test failures due to  System.InvalidOperationException : Memory pool with active blocks is being disposed, 302 of 303 returned

### ISSUE THREAD

**Issue by rynowak** [dotnet/aspnetcore#13030]
I can pretty regularly reproduce test failures running the components tests with this message by running the tests in the components solution under the debugger. 

```txt
 Microsoft.AspNetCore.Components.E2ETest.ServerExecutionTests.LocalizationTest.CanSetCultureAndReadLocalizedResources(culture: "en-US", message: "Hello!")
   Source: LocalizationTest.cs line 38

Outcomes
   1 Passed
   1 Failed

Results

    1)  Microsoft.AspNetCore.Components.E2ETest.ServerExecutionTests.LocalizationTest.CanSetCultureAndReadLocalizedResources(culture: "en-US", message: "Hello!")
      Duration: 29 sec

    2)  Microsoft.AspNetCore.Components.E2ETest.ServerExecutionTests.LocalizationTest.CanSetCultureAndReadLocalizedResources(culture: "en-US", message: "Hello!")
      Duration: 1 ms

      Message: 
        [Test Class Cleanup Failure (Microsoft.AspNetCore.Components.E2ETest.ServerExecutionTests.LocalizationTest)]: System.InvalidOperationException : Memory pool with active blocks is being disposed, 302 of 303 returned
      Stack Trace: 
        MemoryPoolThrowHelper.ThrowInvalidOperationException_DisposingPoolWithActiveBlocks(Int32 returned, Int32 total, DiagnosticPoolBlock[] blocks) line 62
        DiagnosticMemoryPool.Dispose(Boolean disposing) line 119
        MemoryPool`1.Dispose()
        SocketConnectionListener.DisposeAsync() line 144
        KestrelServer.StopAsync(CancellationToken cancellationToken) line 190
        KestrelServer.StopAsync(CancellationToken cancellationToken) line 162
        KestrelServer.Dispose() line 211
        ServiceProviderEngineScope.DisposeAsync()
        --- End of stack trace from previous location where exception was thrown ---
        WebHost.DisposeServiceProviderAsync(IServiceProvider serviceProvider) line 384
        WebHost.DisposeAsync() line 375
        WebHost.Dispose() line 358
        WebHostServerFixture.Dispose() line 27
        ToggleExecutionModeServerFixture`1.Dispose() line 61

```

**Comment by rynowak** [comment]
@halter73 - any ideas?

**Comment by jkotalik** [comment]
_big yikes_ How recently have you started seeing this? I think it could be from my change ~ a week ago that introduced a new type for reading the request body.

**Comment by halter73** [comment]
Seems to be a shutdown bug. Usually this means that the server isn't properly waiting on all the request tasks during shutdown. There's probably still some request logic still running that hasn't returned all it's blocks when the server disposed the memory pool during shutdown. @jkotalik is investigating as we speak.

**Comment by jkotalik** [comment]
Still working on reproing this with more information. I've tried to enable rent tracking on the debug builds but that caused this issue not to repro. I'm now trying to figure out if the rent/return issue on shutdown is on the reader or writer side of the transport.

**Comment by jkotalik** [comment]
Running this in a loop overnight, we shall see.... <_<

**Comment by rynowak** [comment]
It's been at least one night.

**Comment by jkotalik** [comment]
TL;DR I wasn't able to repro it with any additional information/logging, @halter73 started running it on his machine and wasn't able to repro it either... @halter73 should I continue my investigation on this?

**Comment by halter73** [comment]
We should try to get a reliable repro of this. I doubt this will be a ship blocker since it seems shutdown-related, but it does imply some task isn't being observed.

This isn't even necessarily a Kestrel issue (though it certainly could be). An app that does unobserved operations on the request or response body (e.g. `_ = httpContext.Request.Body.CopyToAsync(Stream.Null);`) could lead to the DiagnosticMemoryPool reporting this kind of error during shutdown.

Ultimately, we need to run automated tests on debug builds on a regular basis so we can collect more data. I'll look at getting https://github.com/aspnet/AspNetCore/pull/12292 running on a schedule instead of per-PR so I can merge it.

**Comment by jkotalik** [comment]
Found another one that looks a bit more concerning: https://dev.azure.com/dnceng/public/_build/results?buildId=440892&view=ms.vss-test-web.build-test-results-tab&runId=14042650&resultId=114758&paneView=debug.

```
System.AggregateException : Exceptions occurred while accessing blocks (Block is backed by disposed slab)\n---- System.InvalidOperationException : Block is backed by disposed slab


   at System.Buffers.DiagnosticMemoryPool.WhenAllBlocksReturnedAsync(TimeSpan timeout) in /_/src/Shared/Buffers.MemoryPool/DiagnosticMemoryPool.cs:line 162
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.MaxRequestBufferSizeTests.ServerShutsDownGracefullyWhenMaxRequestBufferSizeExceeded() in /_/src/Servers/Kestrel/test/FunctionalTests/MaxRequestBufferSizeTests.cs:line 281
--- End of stack trace from previous location ---
----- Inner Stack Trace -----
   at System.Buffers.MemoryPoolThrowHelper.ThrowInvalidOperationException_BlockIsBackedByDisposedSlab(DiagnosticPoolBlock block) in /_/src/Shared/Buffers.MemoryPool/MemoryPoolThrowHelper.cs:line 57
   at System.Buffers.DiagnosticPoolBlock.GetSpan() in /_/src/Shared/Buffers.MemoryPool/DiagnosticPoolBlock.cs:line 115
   at System.Buffers.BuffersExtensions.CopyToMultiSegment[T](ReadOnlySequence`1& sequence, Span`1 destination)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 destination, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.MaxRequestBufferSizeTests.<>c__DisplayClass6_0.<<StartWebHost>b__3>d.MoveNext() in /_/src/Servers/Kestrel/test/FunctionalTests/MaxRequestBufferSizeTests.cs:line 330
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application) in /_/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:line 630

```

**Comment by halter73** [comment]
Here's the full logs for the DiagnosticMemoryPool failure:

```
[0.001s] [TestLifetime] [Information] Starting test ServerShutsDownGracefullyWhenMaxRequestBufferSizeExceeded at 2019-11-27T02:09:50
[0.023s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Debug] Hosting starting
[0.030s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Debug] Hosting started
[0.030s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Debug] Loaded hosting startup assembly Sockets.FunctionalTests
[0.111s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Connection id "0HLRIRRHBERBJ" paused.
[0.112s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "0HLRIRRHBERBJ" accepted.
[0.228s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Debug] Hosting shutdown
[0.234s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "0HLRIRRHBERBJ" started.
[0.274s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Information] Request starting HTTP/1.1 POST http:/// - 104857600
[5.288s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Some connections failed to close gracefully during server shutdown.
[5.290s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Connection id "0HLRIRRHBERBJ" sending FIN because: "The connection was aborted because the server is shutting down and request processing didn't complete within the time specified by HostOptions.ShutdownTimeout."
[6.576s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Some connections failed to abort during server shutdown.
[6.580s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "0HLRIRRHBERBJ", Request id "0HLRIRRHBERBJ:00000001": started reading request body.
[6.587s] [Microsoft.AspNetCore.Server.Kestrel] [Error] Connection id "0HLRIRRHBERBJ", Request id "0HLRIRRHBERBJ:00000001": An unhandled exception was thrown by the application.
System.InvalidOperationException: Block is backed by disposed slab
   at System.Buffers.MemoryPoolThrowHelper.ThrowInvalidOperationException_BlockIsBackedByDisposedSlab(DiagnosticPoolBlock block) in /_/src/Shared/Buffers.MemoryPool/MemoryPoolThrowHelper.cs:line 57
   at System.Buffers.DiagnosticPoolBlock.GetSpan() in /_/src/Shared/Buffers.MemoryPool/DiagnosticPoolBlock.cs:line 115
   at System.Buffers.BuffersExtensions.CopyToMultiSegment[T](ReadOnlySequence`1& sequence, Span`1 destination)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 destination, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.MaxRequestBufferSizeTests.<>c__DisplayClass6_0.<<StartWebHost>b__3>d.MoveNext() in /_/src/Servers/Kestrel/test/FunctionalTests/MaxRequestBufferSizeTests.cs:line 330
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application) in /_/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:line 630
[6.589s] [Microsoft.AspNetCore.Server.Kestrel] [Warning] Connection processing ended abnormally.
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'MemoryPool'.
   at System.Buffers.MemoryPoolThrowHelper.ThrowObjectDisposedException(ExceptionArgument argument) in /_/src/Shared/Buffers.MemoryPool/MemoryPoolThrowHelper.cs:line 94
   at System.Buffers.DiagnosticMemoryPool.Rent(Int32 size) in /_/src/Shared/Buffers.MemoryPool/DiagnosticMemoryPool.cs:line 58
   at System.IO.Pipelines.Pipe.AllocateSegment(Int32 sizeHint)
   at System.IO.Pipelines.Pipe.AllocateWriteHeadSynchronized(Int32 sizeHint)
   at System.IO.Pipelines.Pipe.GetSpan(Int32 sizeHint)
   at System.IO.Pipelines.Pipe.DefaultPipeWriter.GetSpan(Int32 sizeHint)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.PipeWriterHelpers.ConcurrentPipeWriter.GetSpan(Int32 sizeHint) in /_/src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs:line 77
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1OutputProducer.WriteResponseHeaders(Int32 statusCode, String reasonPhrase, HttpResponseHeaders responseHeaders, Boolean autoChunk, Boolean appComplete) in /_/src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs:line 344
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProduceStart(Boolean appCompleted) in /_/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:line 994
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProduceEnd() in /_/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:line 1054
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application) in /_/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:line 692
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application) in /_/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:line 546
[6.592s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "0HLRIRRHBERBJ" disconnecting.
[6.592s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "0HLRIRRHBERBJ" stopped.
[6.629s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets] [Debug] Connection id "0HLRIRRHBERBJ" resumed.
[6.615s] [TestLifetime] [Information] Finished test ServerShutsDownGracefullyWhenMaxRequestBufferSizeExceeded in 6.6143044s
```

The log I'm really interested in is "[6.576s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Some connections failed to abort during server shutdown." This is logged [here](https://github.com/aspnet/AspNetCore/blob/3b7cdc166aa3c9733fa60b56d91fc0fff9b11652/src/Servers/Kestrel/Core/src/KestrelServer.cs#L212). In this case the transports and therefore the memory pool is disposed despite not all the requests being completed. In some sense this is expected, we just need to see if there's something about the ServerShutsDownGracefullyWhenMaxRequestBufferSizeExceeded test that causes the middleware not to terminate within the timeout despite the underlying connection being aborted.

**Comment by anurse** [comment]
Pre-planning notes: I feel like this isn't something we can spend resources investigating right now, unless it's actively blocking builds or there's a customer-facing issue identified.

**Comment by anurse** [comment]
We need to run tests on `DEBUG` builds to get more data here, since that's the only place the `DiagnosticsMemoryPool` is used.

**Comment by kostebudinoski** [comment]
It happened on production for a solution we run on asp.net core 3.1:

```
2020-07-08 14:20:41.064 +00:00 [Error] [Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware] An exception was thrown attempting to display the error page.
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'MemoryPool'.
   at System.Buffers.MemoryPoolThrowHelper.ThrowObjectDisposedException(ExceptionArgument argument)
   at System.Buffers.SlabMemoryPool.Lease()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1OutputProducer.GetSpan(Int32 sizeHint)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpResponsePipeWriter.GetSpan(Int32 sizeHint)
   at Microsoft.AspNetCore.Http.HttpResponseWritingExtensions.Write(HttpResponse response, String text, Encoding encoding)
   at Microsoft.AspNetCore.Http.HttpResponseWritingExtensions.WriteAsync(HttpResponse response, String text, Encoding encoding, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Http.HttpResponseWritingExtensions.WriteAsync(HttpResponse response, String text, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.DisplayException(ErrorContext errorContext)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by halter73** [comment]
> It happened on production
> ...
> ```   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.DisplayException(ErrorContext errorContext)```

😨

**Comment by halter73** [comment]
Seriously though. There's a warning [here in our docs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-3.1#developer-exception-page) against enabling the developer exception page in production because it can potentially leak sensitive data.

DeveloperExceptionPageMiddleware.Invoke being at the bottom of the stack, probably explains the ObjectDisposedException. It looks like the call to DeveloperExceptionPageMiddleware.Invoke(HttpContext) is not being being awaited from wherever it's being invoked. As I noted earlier:

> This isn't even necessarily a Kestrel issue (though it certainly could be). An app that does unobserved operations on the request or response body (e.g. _ = httpContext.Request.Body.CopyToAsync(Stream.Null);) could lead to the DiagnosticMemoryPool reporting this kind of error during shutdown.

**Comment by kostebudinoski** [comment]
There is a check for environment for the developer exception page, but there was a deployment bug affecting the aspnetcore environment. Thanks for noticing this! Will monitor further!

**Comment by sh-erfan** [comment]
I get this when syncing my database(about 1GB) through api,  using Http client, one request per 30 seconds. Each request returns about 5Mb of data. App pool shuts down after a while.
This is the log:

```
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'MemoryPool'.
   at System.Buffers.MemoryPoolThrowHelper.ThrowObjectDisposedException(ExceptionArgument argument)
   at System.Buffers.SlabMemoryPool.Lease()
   at System.Buffers.SlabMemoryPool.Rent(Int32 size)
   at System.IO.Pipelines.Pipe.AllocateSegment(Int32 sizeHint)
   at System.IO.Pipelines.Pipe.AllocateWriteHeadSynchronized(Int32 sizeHint)
   at Microsoft.AspNetCore.Server.IIS.Core.OutputProducer.WriteAsync(ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.<>c__DisplayClass335_0.<<WriteAsync>g__WriteFirstAsync|0>d.MoveNext()
```

**Comment by halter73** [comment]
@sh-erfan Do you have any logs? This is expected behavior if an ongoing request takes longer than the [ShutdownTimeout](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-5.0#shutdowntimeout) (defaults to 5 seconds). You can use [IHostApplicationLifetime](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-5.0#ihostapplicationlifetime).ApplicationStopping to observe this and cancel the request gracefully before the timeout.

Even after the ShutdownTimeout expires, Kestrel gives another (nonconfigurable) 1 second grace period after closing all its sockets before it disposes the memory pool. In many cases this is enough time for middleware to observe an IOException and exit, but that is definitely not guaranteed to work. For example, middleware doing some operation that takes over 6 seconds before first writing to a response or middleware delaying over a second between writing without observing ApplicationStopping could run into this MemoryPool ODE.

**Comment by sh-erfan** [comment]
> @sh-erfan Do you have any logs? This is expected behavior if an ongoing request takes longer than the [ShutdownTimeout](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-5.0#shutdowntimeout) (defaults to 5 seconds). You can use [IHostApplicationLifetime](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-5.0#ihostapplicationlifetime).ApplicationStopping to observe this and cancel the request gracefully before the timeout.
> 
> Even after the ShutdownTimeout expires, Kestrel gives another (nonconfigurable) 1 second grace period after closing all its sockets before it disposes the memory pool. In many cases this is enough time for middleware to observe an IOException and exit, but that is definitely not guaranteed to work. For example, middleware doing some operation that takes over 6 seconds before first writing to a response or middleware delaying over a second between writing without observing ApplicationStopping could run into this MemoryPool ODE.

@halter73 Absolutely correct. The reason was default ShutdownTimeout. Thanks.

**Comment by adityamandaleeka** [comment]
Closing for now... let's see if this happens again.
